### PR TITLE
Check class initialization status in reusing iprofiling data and inlined sites relocation

### PR DESCRIPTION
**Pull request by liqunl**
Profiled class from last run might not get used in the second run. If
we don't check its initialization status, the profiling data might
be inconsistent with the class hierarchy, resulting in bad inlining
decision or even wrong optimization.

This PR also disable activation of method whose class hasn't been
initialized, except for interface classes, as interface classes are
special in terms of initialization.

Fixes: #11179

Signed-off-by: Liqun Liu <liqunl@ca.ibm.com>

### CODE REVIEW

**Comment by liqunl** [comment]
@dsouzai I wonder if we should throw away the whole `_csInfo` from old iprofiler data if one of the class is invalid? If we keep copy others, it'll be inaccurate.

**Comment by dsouzai** [comment]
On the one hand, the if the profiling data isn't correct because we've removed one of the classes in the entry, then we may inline a less frequent implementation which isn't necessarily any worse than having no information. On the other hand though, we may never get a chance to correct the entry because there might not enough interpreter profiling data.

So, maybe it's best you just modify the two `TR_IProfiler::persistentProfilingSample` to check if the entry is a `TR_IPBCDataCallGraph`, in which case you can go through and check if all the ram classes are initialized; if they aren't then you can just return NULL. This way, the next time the compiler asks for this entry, it can still look it up in the SCC and then decide to merge the data with the current entry in the IProfiler's hash table.

What do you think @mpirvu ?

Also, @liqunl, since you're opening this PR, could you also added the following change:
```
diff --git a/runtime/compiler/runtime/RelocationRecord.cpp b/runtime/compiler/runtime/RelocationRecord.cpp
index 18214f20e5..680c7be697 100644
--- a/runtime/compiler/runtime/RelocationRecord.cpp
+++ b/runtime/compiler/runtime/RelocationRecord.cpp
@@ -2284,6 +2284,9 @@ TR_RelocationRecordInlinedMethod::inlinedSiteCanBeActivated(TR_RelocationRuntime
       return false;
       }
 
+   if (!reloRuntime->fej9()->isClassInitialized((TR_OpaqueClassBlock*)J9_CLASS_FROM_METHOD(currentMethod)))
+      return false;
+
    return true;
    }
```
Basically at relo time, we check if the inlined method's class is initialized.

**Comment by liqunl** [comment]
@dsouzai In terms of the initialization check on class of inlined method, it might not work on interface class. I have an impression that an interface class may never be initialized because initialization happens at accessing static fields, calling static methods and instantiating the class, and interface may not have static members and can't be instantiated. However, the default method of an interface can be inlined. @tajila Could you shed some light on this?

**Comment by liqunl** [comment]
> On the one hand, the if the profiling data isn't correct because we've removed one of the classes in the entry, then we may inline a less frequent implementation which isn't necessarily any worse than having no information. On the other hand though, we may never get a chance to correct the entry because there might not enough interpreter profiling data.

This makes a good point
> So, maybe it's best you just modify the two TR_IProfiler::persistentProfilingSample to check if the entry is a TR_IPBCDataCallGraph, in which case you can go through and check if all the ram classes are initialized; if they aren't then you can just return NULL. This way, the next time the compiler asks for this entry, it can still look it up in the SCC and then decide to merge the data with the current entry in the IProfiler's hash table.

Do we merge old profiling data with new one from current run as mentioned by Marius in https://github.com/eclipse/openj9/issues/11179#issuecomment-739062030? If so, I think it's fine to leave partial profiling data from last run and allow it to merge with data from current run.

**Comment by tajila** [comment]
> However, the default method of an interface can be inlined. @tajila Could you shed some light on this?

There are a few ways an interface can be initialized:

```
The execution of any one of the Java Virtual Machine instructions new, getstatic, putstatic, or invokestatic that references C (§new, §getstatic, §putstatic, §invokestatic).

Upon execution of a new instruction, the class to be initialized is the class referenced by the instruction.

Upon execution of a getstatic, putstatic, or invokestatic instruction, the class or interface to be initialized is the class or interface that declares the resolved field or method. 
```

Same goes for methodhavle operations, as well as:

```
If C is an interface that declares a non-abstract, non-static method, the initialization of a class that implements C directly or indirectly.
```

So as long as the implementing class that exposes the default method (ie. non-abstract, non-static method) is initialized than the the interface will be initialized.

**Comment by dsouzai** [comment]
> Do we merge old profiling data with new one from current run as mentioned by Marius in #11179 (comment)? If so, I think it's fine to leave partial profiling data from last run and allow it to merge with data from current run.

Doesn't look like we merge profiling data; we either use the one from the SCC or the data collected in the current run:
https://github.com/eclipse/openj9/blob/c60b68ad083fb64d7936759202ce48deb52c6ed2/runtime/compiler/runtime/IProfiler.cpp#L1547-L1563

**Comment by liqunl** [comment]
@dsouzai From an offline discussion with @tajila, the instantiation of implementers of interface with default method will initialize the interface class, interface classes without static fields/methods nor default method won't be initialized except when reflection operation applies to them (such as `Class.getName()`). 
So if a default method is inlined, class of this method is an interface. If the interface hasn't been initialized, this means none of its implementers have been instantiated. Is invalidating the inlined site what we want in this scenario?

>Doesn't look like we merge profiling data; we either use the one from the SCC or the data collected in the current run:

Shouldn't we always choose the most recent one?

**Comment by mpirvu** [comment]
> Doesn't look like we merge profiling data; we either use the one from the SCC or the data collected in the current run

The merging happens later. Once we decide to pick info from SCC, we will load that information into an entry in the IProfiler hashtable. That entry continues to receive samples (until the method is compiled) and be updated with information from the current run.

**Comment by dsouzai** [comment]
> So if a default method is inlined, class of this method is an interface. If the interface hasn't been initialized, this means none of its implementers have been instantiated. Is invalidating the inlined site what we want in this scenario?

Ok, then maybe the check should be something like
```
TR_OpaqueClassBlock *classOfMethod = (TR_OpaqueClassBlock*)J9_CLASS_FROM_METHOD(currentMethod);
if (!reloRuntime->fej9()->isInterfaceClass(classOfMethod) && !reloRuntime->fej9()->isClassInitialized(classOfMethod))
   return false;
```
The reason being:
* If the interface class doesn't have any default methods or static fields/methods, then it's not going to get initialized at all so no point checking at relo time.
* If the interface class **does** have default methods or static fields/methods, then even if it isn't initialized at relo time, by the time we execute the inlined code, it will have been initialized because instantiating an implementer will have initialized it, so again no point checking at relo time.

**Comment by dsouzai** [comment]
> Shouldn't we always choose the most recent one?

Recent is hard to characterize when you consider that one entry is coming from a previous run. The quality of the data is based on how many samples it has.

**Comment by liqunl** [comment]
@dsouzai Based on Mariu's comment https://github.com/eclipse/openj9/pull/11373#issuecomment-740010725, I keep the original fix in iprofiling data loading. I added the change on inlined site relocation as you suggested, could you review again?

File: runtime/compiler/runtime/RelocationRecord.cpp [hunk 1]
```diff
@@ -2272,6 +2272,13 @@ TR_RelocationRecordInlinedMethod::inlinedSiteCanBeActivated(TR_RelocationRuntime
       return false;
       }
 
+   TR_OpaqueClassBlock *classOfMethod = (TR_OpaqueClassBlock*)J9_CLASS_FROM_METHOD(currentMethod);
+   if (!reloRuntime->fej9()->isInterfaceClass(classOfMethod) && !reloRuntime->fej9()->isClassInitialized(classOfMethod))
```
Comment (dsouzai) [inline runtime/compiler/runtime/RelocationRecord.cpp:2276]: Could you add a comment here describing why we're excluding interface classes.

File: runtime/compiler/runtime/IProfiler.cpp [hunk 2]
```diff
@@ -3153,7 +3153,7 @@ TR_IPBCDataCallGraph::loadFromPersistentCopy(TR_IPBCDataStorageHeader * storage,
          if (comp->fej9()->sharedCache()->isROMClassOffsetInSharedCache(csInfoClazzOffset, &romClass))
             ramClass = ((TR_J9VM *)comp->fej9())->matchRAMclassFromROMclass((J9ROMClass *)romClass, comp);
 
-         if (ramClass)
+         if (ramClass && comp->fej9()->isClassInitialized((TR_OpaqueClassBlock*)ramClass))
```
Comment (dsouzai) [inline runtime/compiler/runtime/IProfiler.cpp:3156]: Might be worth adding a comment here about why this is necessary (basically what it results in the optimizer) and also why we're ok with having the ith entry be NULL (essentially just a brief summary of Marius' comments).

**Review by dsouzai — CHANGES_REQUESTED** [review]
Functionally looks good to me, just requesting comments.

**Comment by dsouzai** [comment]
Do you also mind doing run of something with AOT with and without your change? Worth knowing beforehand if this change would dramatically increase load failures.

**Review by mpirvu — COMMENTED** [review]
I am ok with the change. It can done in different ways, but it's hard to tell which variant is better.
So, when asking about IP info for a particular pc for the first time we consult two pieces of information:
1) IP information collected in this run (if any, because IProfiler is turned off during start-up phase)
2) IP information stored in the SCC
If information exists in one place but not in the other, we just return that. Note that if we pick SCC as the source, the entry is copied from the SCC into the runtime IProfiler hashtable where will continue to collect IProfiler samples until the method it belongs to get compiled.
If the information exists in both places, we pick the entry with most samples (we first create a dummy IP entry on stack, copy information from SCC into this dummy entry - with `loadFromPersistentCopy()` -, do the comparison with the entry in the IProfiler hashtable - based on `getSamplingCount()`  and then possibly overwrite the entry in the IProfiler hashtable with the values from the dummy entry).

One variant could be to totally ignore the SCC entry if the class with the most samples is not initialized, on the account that the SCC profile information is not representative for the current run.

**Review by mpirvu — APPROVED** [review]

**Review by dsouzai — APPROVED** [review]

**Comment by dsouzai** [comment]
jenkins test sanity xlinux jdk11

**Comment by liqunl** [comment]
Removed the inlined site allocation change after an offline discussion with Irwin. The change is not needed because the initialization check is done in constant pool entry relocation. For non-interface methods, there must be a reference of the class in the constant pool, successful relocation of that class implies initialization status. For class of default method, which is an interface, the relocation doesn't guarantee initialization, but execution of default method require an instance to be created, which will initialize the class, so we're good.
